File: core/src/main/kotlin/fr/rob/core/network/v2/netty/protobuf/ProtobufNettyChannelInitializer.kt
package fr.rob.core.network.v2.netty.protobuf

import com.google.protobuf.MessageLite
import fr.rob.core.network.v2.ServerInterface
import fr.rob.core.network.v2.netty.NettyChannelInitializer
import io.netty.channel.ChannelPipeline
import io.netty.handler.codec.LengthFieldBasedFrameDecoder
import io.netty.handler.codec.LengthFieldPrepender
import io.netty.handler.codec.protobuf.ProtobufDecoder
import io.netty.handler.codec.protobuf.ProtobufEncoder

class ProtobufNettyChannelInitializer<T>(private val prototype: MessageLite, private val server: ServerInterface<T>, ssl: Boolean = false) :
    NettyChannelInitializer<T>(ssl) {

    override fun channelHandler(): fr.rob.core.network.v2.netty.NettyChannelHandler<T> =
        ProtobufNettyChannelHandler(server)

    override fun registerHandlers(pipeline: ChannelPipeline) {
        // Decoders
        pipeline.addLast(
            "frameDecoder",
            LengthFieldBasedFrameDecoder(1048576, 0, 4, 0, 4)
        )
        pipeline.addLast(
            "protobufDecoder",
            ProtobufDecoder(prototype)
        )

        // Encoders
        pipeline.addLast("frameEncoder", LengthFieldPrepender(4))
        pipeline.addLast("protobufEncoder", ProtobufEncoder())
    }
}

File: RobClient/src/network/ProtobufNettyChannelInitializer.cs
using DotNetty.Codecs;
using DotNetty.Codecs.Protobuf;
using DotNetty.Transport.Channels;
using DotNetty.Transport.Channels.Sockets;
using Fr.Raven.Proto.Message.Gateway;

namespace RobClient.Network {
    class ProtobufNettyChannelInitializer : ChannelInitializer<ISocketChannel>
    {
        private Client _client;

        public ProtobufNettyChannelInitializer(Client client) {
            _client = client;
        }

        protected override void InitChannel(ISocketChannel channel)
        {
            IChannelPipeline pipeline = channel.Pipeline;
            pipeline.AddLast("frameEncoder", new LengthFieldPrepender(4));
            pipeline.AddLast("frameDecoder", new LengthFieldBasedFrameDecoder(1048576, 0, 4, 0, 4));

            pipeline.AddLast("protobufEncoder", new ProtobufEncoder());
            pipeline.AddLast("protobufDecoder", new ProtobufDecoder(Packet.Parser));

            pipeline.AddLast(new NettyChannelHandler(_client));
        }
    }
}
